### PR TITLE
Show loan position

### PR DESCRIPTION
**Pull request by graykode**

### CODE REVIEW

File: api/loan-position.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    },
+  )
+  return loanPositions.map((loanPosition) => ({
+    positionId: BigInt(loanPosition.id),
+    substitute: toCurrency(loanPosition.substitute),
+    underlying: toCurrency(loanPosition.underlying),
+    collateral: {
+      underlying: toCurrency(loanPosition.collateral.underlying),
+      substitute: toCurrency(loanPosition.collateral.substitute),
+      liquidationThreshold: loanPosition.collateral.liquidationThreshold,
```
Comment (onemonster) [inline api/loan-position.ts:26]: I think BigInt is appropriate here, even though the number is always safe. Numbers that use decimals precision should be `bigint`

File: api/loan-position.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      liquidationTargetLtv: loanPosition.collateral.liquidationTargetLtv,
```
Comment (onemonster) [inline api/loan-position.ts:27]: (2)

File: api/loan-position.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    },
+    interest: BigInt(loanPosition.amount) - BigInt(loanPosition.principal),
+    amount: BigInt(loanPosition.amount),
+    collateralAmount: BigInt(loanPosition.collateralAmount),
+    expiryEpoch: Number(loanPosition.toEpoch.id),
+    expiryTimestamp: Number(loanPosition.toEpoch.endTimestamp),
+    ltv: Number(loanPosition.ltv),
```
Comment (onemonster) [inline api/loan-position.ts:34]: Is this number decimal (23.34%) or bigint (233400). If it's the latter, we should use `BigInt` type

File: components/borrow.tsx [hunk 4]
```diff
@@ -48,56 +51,55 @@ const EditSvg = (props: SVGProps<any>) => (
...
+              {position.underlying.name}
+            </div>
           </div>
         </div>
         <div className="flex flex-col items-end">
-          <div className="font-bold">{apy}</div>
+          <div className="font-bold">
+            {calculateApr(
+              Number(position.interest) / Number(position.amount),
+              position.expiryTimestamp - currentTimestamp,
```
Comment (onemonster) [inline components/borrow.tsx:92]: You must use the timestamp the borrow was initiated, or the APR won't be fixed.

File: components/borrow.tsx [hunk 5]
```diff
@@ -126,12 +145,14 @@ const Position = ({
...
-            <div className="flex text-green-500 text-xs sm:text-sm">{ltv}</div>
+            <div className="flex text-green-500 text-xs sm:text-sm">
+              {position.ltv.toFixed(2)}%
+            </div>
           </div>
           <div className="flex items-center justify-between text-xs">
             <div className="text-gray-500">Liquidation Threshold</div>
             <div className="flex text-xs sm:text-sm">
-              {liquidationThreshold}
+              {Number(position.collateral.liquidationThreshold) / 10000}%
```
Comment (onemonster) [inline components/borrow.tsx:155]: Let's use `formatUnits` here instead.

File: components/borrow.tsx [hunk 6]
```diff
@@ -188,11 +209,13 @@ const Asset = ({
               <div className="text-gray-500 text-xs">{currency.name}</div>
             </div>
           </div>
-          <div className="text-sm font-bold sm:w-[80px]">{apy.toFixed(2)}</div>
+          <ClientComponent className="text-sm font-bold sm:w-[80px]">
+            {apy.toFixed(2)}
```
Comment (onemonster) [inline components/borrow.tsx:213]: Should this be `apr`?

File: components/borrow.tsx [hunk 7]
```diff
@@ -273,66 +296,46 @@ const Borrow = ({
...
               <div className="font-bold">
                 $
-                {positions.reduce(
-                  (acc, position) => +position.price * +position.borrowed + acc,
-                  0,
-                )}
+                {positions
+                  .reduce(
+                    (acc, { underlying, amount }) =>
+                      +formatUnits(amount, underlying.decimals) *
```
Comment (onemonster) [inline components/borrow.tsx:302]: Let's use the `dollarValue` function to do the math here.

File: components/borrow.tsx [hunk 8]
```diff
@@ -368,9 +371,9 @@ const Borrow = ({
                 <Asset
                   key={i}
                   currency={assetStatus.underlying}
-                  apy={apy[assetStatus.underlying.address] ?? 0}
-                  available={available[assetStatus.underlying.address] ?? 0n}
-                  borrowed={borrowed[assetStatus.underlying.address] ?? 0n}
+                  apy={12.34 ?? 0}
+                  available={100000000000n ?? 0n}
```
Comment (onemonster) [inline components/borrow.tsx:375]: Will this be handled in a different PR?

File: components/borrow.tsx (hunk 7 above)
Comment (graykode) [reply components/borrow.tsx]: https://github.com/clober-dex/coupon.finance/pull/77/commits/4545c84ab98b9e4f5a2bced9147d9e6c83a3a060

File: api/loan-position.ts (hunk 1 above)
Comment (graykode) [reply api/loan-position.ts:26]: @onemonster, no, it can't be because Asset is used in getServerSideProps.

File: components/borrow.tsx (hunk 8 above)
Comment (graykode) [reply components/borrow.tsx]: https://github.com/clober-dex/coupon.finance/pull/77/commits/c6a55b42da247d3711b55140318147ca06a3e413 @onemonster Good

File: components/borrow.tsx [hunk 9]
```diff
@@ -273,66 +302,50 @@ const Borrow = ({
...
-                }
-                onEditExpiry={() =>
-                  setEditExpiryPosition({
-                    currency: position.currency,
-                    amount: position.borrowed,
-                  })
+                position={position}
+                price={prices[position.underlying.address] ?? 0}
+                collateralPrice={
+                  prices[position.collateral.underlying.address] ?? 0
```
Comment (onemonster) [inline components/borrow.tsx:343]: `remove ?? 0`

File: components/borrow.tsx [hunk 10]
```diff
@@ -371,7 +384,7 @@ const Borrow = ({
                   apy={apy[assetStatus.underlying.address] ?? 0}
                   available={available[assetStatus.underlying.address] ?? 0n}
                   borrowed={borrowed[assetStatus.underlying.address] ?? 0n}
-                  price={prices[assetStatus.underlying.address]}
+                  price={prices[assetStatus.underlying.address] ?? 0}
```
Comment (onemonster) [inline components/borrow.tsx:387]: remove `?? 0`

File: contexts/borrow-context.tsx [hunk 11]
```diff
@@ -214,14 +156,29 @@ export const BorrowProvider = ({ children }: React.PropsWithChildren<{}>) => {
     ],
   )
 
+  // TODO: get apy from order book
```
Comment (onemonster) [inline contexts/borrow-context.tsx:159]: Let's remove these.

Comment (graykode) [reply contexts/borrow-context.tsx]: https://github.com/clober-dex/coupon.finance/pull/77/commits/d9e558e4b45744132c34919f1d7508808ead92aa

File: components/borrow.tsx (hunk 10 above)
Comment (graykode) [reply components/borrow.tsx]: https://github.com/clober-dex/coupon.finance/pull/77/commits/1e21407d5348e1786b1f5788f2d57ee999322246

File: components/borrow.tsx (hunk 9 above)
Comment (graykode) [reply components/borrow.tsx]: [### ](https://github.com/clober-dex/coupon.finance/pull/77/commits/1e21407d5348e1786b1f5788f2d57ee999322246)

File: components/borrow.tsx (hunk 4 above)
Comment (graykode) [reply components/borrow.tsx:98]: Could you also check if this one is correct? : https://github.com/clober-dex/coupon.finance/blob/dev/components/deposit.tsx#L55

File: components/borrow.tsx (hunk 5 above)
Comment (graykode) [reply components/borrow.tsx]: https://github.com/clober-dex/coupon.finance/pull/77/commits/2914777377730460505c808c72903ad698a4bfec Done!

File: components/borrow.tsx (hunk 4 above)
Comment (onemonster) [reply components/borrow.tsx:98]: Wow they're both wrong lol

Comment (onemonster) [reply components/borrow.tsx:98]: They're both wrong. Once a position is made the apr, apy does not change

Comment (graykode) [reply components/borrow.tsx]: https://github.com/clober-dex/coupon.finance/pull/77/commits/315f0be921174528ad374363e9636ea6faa5f2de

@onemonster Could you check this commit?

File: api/loan-position.ts (hunk 1 above)
Comment (onemonster) [reply api/loan-position.ts:26]: Agreed but doesn't `toCurrency` make use of `bigint` as well? Just wondering why currency works and bigint doesn't.

**Review by onemonster — DISMISSED** [review]
